File: algorithm/leetcodeObjectiveC/kth-largest-element-in-an-array.m
// 215. Kth Largest Element in an Array
// Difficulty: Medium

// Find the kth largest element in an unsorted array.

// Note that it is the kth largest element in the sorted order, not the kth distinct element.

// For example,
// Given [3,2,1,5,6,4] and k = 2, return 5.

// Note:
// You may assume k is always valid, 1 ≤ k ≤ array length.

// Time:  O(n) ~ O(n^2)
// Space: O(1)

#import <Foundation/Foundation.h>

NSInteger PartitionAroundPivot(NSInteger left, NSInteger right, NSInteger pivotIdx, NSMutableArray* nums) {
  [nums exchangeObjectAtIndex:pivotIdx withObjectAtIndex:right];
  int pivot = [nums[right] intValue];
  NSInteger newPivotIdx = left;
  for (NSInteger i = left; i < right; ++i) {
    if ([nums[i] intValue] > pivot) {
      [nums exchangeObjectAtIndex:newPivotIdx++ withObjectAtIndex:i];
    }
  }
  [nums exchangeObjectAtIndex:newPivotIdx withObjectAtIndex:right];
  return newPivotIdx;
}

int findKthLargest(NSMutableArray* nums, int k) {
  NSInteger left = 0, right = [nums count] - 1;
  while (left <= right) {
    NSInteger pivotIdx = left + rand() % (right - left + 1);
    NSInteger newPivotIdx = PartitionAroundPivot(left, right, pivotIdx, nums);
    if (newPivotIdx == k - 1) {
      return [nums[newPivotIdx] intValue];
    } else if (newPivotIdx > k - 1) {
      right = newPivotIdx - 1;
    } else {
      left = newPivotIdx + 1;
    }
  }
  return [nums[left] intValue];
}

File: data-structure-and-algorithm/typescript/leetcode/src/kth-largest-element-in-an-array.ts
/*
Given an integer array nums and an integer k, return the kth largest element in the array.

Note that it is the kth largest element in the sorted order, not the kth distinct element.



Example 1:

Input: nums = [3,2,1,5,6,4], k = 2
Output: 5
Example 2:

Input: nums = [3,2,3,1,2,4,5,5,6], k = 4
Output: 4


Constraints:

1 <= k <= nums.length <= 104
-104 <= nums[i] <= 104
*/

function findKthLargest(nums: number[], kth: number): number {
  return findKth((nums, a, b) => nums[b] - nums[a]);

  function findKth(
    compare: (nums: number[], a: number, b: number) => number
  ): number {
    let left = 0;
    let right = nums.length - 1;
    while (left <= right) {
      const randomIndex = left + Math.trunc(Math.random() * (right - left + 1));
      const partitionIndex = partition(nums, left, right, randomIndex);

      if (partitionIndex === kth - 1) {
        return nums[kth - 1];
      } else if (partitionIndex > kth - 1) {
        right = partitionIndex - 1;
      } else {
        left = partitionIndex + 1;
      }
    }

    return nums[left];

    function partition(
      items: number[],
      left: number,
      right: number,
      pivotIndex: number
    ): number {
      swap(items, right, pivotIndex);

      let nextPivotIndex = left;
      for (let index = left; index < right; ++index) {
        if (compare(items, index, right) < 0) {
          swap(items, nextPivotIndex++, index);
        }
      }

      swap(items, nextPivotIndex, right);
      return nextPivotIndex;
    }

    function swap(items: number[], i: number, j: number): void {
      const temp = items[i];
      items[i] = items[j];
      items[j] = temp;
    }
  }
}
